ts/hooks/use_wallet.ts

```diff
@@ -0,0 +1,105 @@
+
+import { asyncDispatcher } from 'ts/redux/async_dispatcher';
+import { Dispatcher } from 'ts/redux/dispatcher';
+import { State } from 'ts/redux/reducer';
+import { analytics } from 'ts/utils/analytics';
+import { constants } from 'ts/utils/constants';
+import { providerStateFactory } from 'ts/utils/providers/provider_state_factory';
+
+const PROVIDER_CHAIN_CHANGED_EVENT = 'chainChanged';
+const PROVIDER_ACCOUNTS_CHANGED_EVENT = 'accountsChanged';
+const PROVIDER_NETWORK_CHANGED_EVENT = 'networkChanged';
+
+export const useWallet = () => {
+    const dispatch = useDispatch();
+    const [dispatcher, setDispatcher] = useState<Dispatcher | undefined>(undefined);
+
+    const accountsChangedRef = useRef<() => void | undefined>(undefined);
+    const networkChangedRef = useRef<(networkId: string) => void | undefined>(undefined);
+
+    const currentNetworkId = useSelector((state: State) => state.networkId);
+    const currentProviderState = useSelector((state: State) => state.providerState);
+
+    useEffect(() => {
+        setDispatcher(new Dispatcher(dispatch));
+    }, [dispatch]);
+
+    const handleAccountsChange = useCallback(async () => {
+        await asyncDispatcher.fetchAccountInfoAndDispatchToStoreAsync(
+            currentProviderState,
+            dispatcher,
+            currentNetworkId,
+            true,
+        );
+    }, [currentNetworkId, currentProviderState, dispatcher]);
+
+    const handleNetworksChange = useCallback(
+        (newNetworkId: string) => {
+            const networkId = parseInt(newNetworkId, 10);
+
+            if (networkId) {
+                dispatcher.updateNetworkId(networkId);
+            }
+        },
+        [dispatcher],
+    );
+
+    const cleanupCurrentProvider = useCallback(() => {
+        const provider = currentProviderState.provider as any;
+
+        if (provider.off) {
+            if (accountsChangedRef.current) {
+                provider.off(PROVIDER_ACCOUNTS_CHANGED_EVENT, accountsChangedRef.current);
+            }
+
+            if (networkChangedRef.current) {
+                provider.off(PROVIDER_NETWORK_CHANGED_EVENT, networkChangedRef.current);
+                provider.off(PROVIDER_CHAIN_CHANGED_EVENT, networkChangedRef.current);
+            }
+        }
+    }, [currentProviderState.provider]);
+
+    const connectToWallet = useCallback(async () => {
+        cleanupCurrentProvider();
+        const providerState = providerStateFactory.getInitialProviderState(currentNetworkId);
+
+        const networkId = await providerState.web3Wrapper.getNetworkIdAsync();
+        if (networkId !== currentNetworkId) {
+            dispatcher.updateNetworkId(networkId);
+        }
+
+        await asyncDispatcher.fetchAccountInfoAndDispatchToStoreAsync(providerState, dispatcher, networkId, true);
+
+        const newProvider = providerState.provider as any;
+        if (newProvider.on) {
+            accountsChangedRef.current = handleAccountsChange;
```

Comment:
Yhis is kinda funky but I couldn't think of a better way to do this in order to preserve the fn reference to be able to unsubscribe in the future. I'm open to suggestions.

Reply:
That's totally fine -- I use useRef as a box like that too (that's pretty standard to hold on to things that you don't care about rendering updates for [that's really the only footgun - useRef assignments don't trigger a rerender)